# Tarefa 03

- Leia os enunciados com atenção
- Saiba que pode haver mais de uma resposta correta
- Insira novas células de código sempre que achar necessário
- Em caso de dúvidas, procure os Tutores
- Divirta-se :)

In [1]:
import pandas as pd
import requests

####  1) Lendo de APIs
Vimos em aula como carregar dados públicos do governo através de um API (*Application Programming Interface*). No exemplo de aula, baixamos os dados de pedidos de verificação de limites (PVL) realizados por estados, e selecionamos apenas aqueles referentes ao estado de São Paulo.

1. Repita os mesmos passos feitos em aula, mas selecione os PVLs realizados por municípios no estado do Rio de Janeiro.
2. Quais são os três *status* das solicitações mais frequentes na base? Quais são suas frequências?
3. Construa uma nova variável que contenha o ano do **status**. Observe que ```data_status``` vem como tipo *object* no **DataFrame**. Dica: você pode usar o método ```.str``` para transformar o tipo da variável em string, em seguida um método como [**slice()**](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.slice.html) ou [**split()**](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.split.html).
4. Indique a frequência de cada ano do campo construído no item (3).

In [2]:
# 1) Seu código aqui
url = 'http://apidatalake.tesouro.gov.br/ords/sadipem/tt/pvl?uf=RJ&tipo_interessado=Munic%C3%ADpio'

In [3]:
r = requests.get(url)

In [4]:
r.status_code

200

In [5]:
data_json = r.json()

In [24]:
df = pd.DataFrame(data_json['items'])
df.head(5)

id_pleito tipo_interessado          interessado  cod_ibge  uf  \
0      65245        Município       Rio de Janeiro   3304557  RJ   
1      66686        Município  São Pedro da Aldeia   3305208  RJ   
2      63101        Município       Rio de Janeiro   3304557  RJ   
3      11005        Município             Araruama   3300209  RJ   
4       7993        Município            Rio Claro   3304409  RJ   

                num_pvl                                             status  \
0  PVL02.001203/2023-06  Encaminhado à PGFN com manifestação técnica fa...   
1  PVL02.003206/2023-76                                           Deferido   
2  PVL02.009023/2022-83  Encaminhado à PGFN com manifestação técnica fa...   
3                  None                     Arquivado por decurso de prazo   
4                  None                                           Deferido   

           num_processo        data_protocolo  \
0  17944.102641/2023-45  2023-07-26T13:58:05Z   
1  17944.104387/2023-10  2023-09-12T15:46:32Z   
2  17944.100196/2023-89  2023-02-06T13:28:28Z   
3  17944.001334/2011-12  2012-12-21T02:00:00Z   
4  17944.000691/2010-74  2011-06-01T03:00:00Z   

                                       tipo_operacao  \
0  Operação contratual interna (com garantia da U...   
1                        Operação contratual interna   
2  Operação contratual interna (com garantia da U...   
3                        Operação contratual interna   
4                        Operação contratual interna   

                                          finalidade  \
0                                     Pró-Transporte   
1                                     Infraestrutura   
2                                     Infraestrutura   
3  PAC 2 - Pró-Transporte - Pavimentação e Qualif...   
4                           Caminho da Escola 2009 A   

                       tipo_credor                   credor moeda  \
0  Instituição Financeira Nacional  Caixa Econômica Federal  Real   
1  Instituição Financeira Nacional      Banco do Brasil S/A  Real   
2  Instituição Financeira Nacional      Banco do Brasil S/A  Real   
3  Instituição Financeira Nacional  Caixa Econômica Federal  Real   
4  Instituição Financeira Nacional      Banco do Brasil S/A  Real   

          valor  pvl_assoc_divida  pvl_contradado_credor data_status  
0  6.459310e+08                 0                      1  03/08/2023  
1  1.200000e+07                 0                      1  19/09/2023  
2  1.200000e+09                 0                      1  16/02/2023  
3  3.359321e+07                 0                      0  09/01/2013  
4  5.845000e+05                 0                      0  22/07/2011

In [8]:
# 2) Seu código aqui
df["status"].value_counts()

status
Deferido                                                 106
Arquivado                                                 53
Regularizado                                              26
Arquivado por decurso de prazo                            24
Deferido (PVL-IF)                                         24
Encaminhado à PGFN com manifestação técnica favorável     16
Arquivado a pedido                                         8
Em retificação pelo interessado                            3
Arquivado pela STN                                         1
Arquivado a pedido (PVL-IF)                                1
Assinado pelo interessado (retificação)                    1
Em retificação pelo credor                                 1
Name: count, dtype: int64

### Os três status com mais frequêcia são: "Deferido", "Arquivado" e "Regularizado" e suas frequências são: 106, 53 e 26 respectivamente

In [9]:
# 3) Seu código aqui
df.dtypes

id_pleito                  int64
tipo_interessado          object
interessado               object
cod_ibge                   int64
uf                        object
num_pvl                   object
status                    object
num_processo              object
data_protocolo            object
tipo_operacao             object
finalidade                object
tipo_credor               object
credor                    object
moeda                     object
valor                    float64
pvl_assoc_divida           int64
pvl_contradado_credor      int64
data_status               object
dtype: object

In [10]:
datas_completas = df['data_status']
datas_completas

0      03/08/2023
1      19/09/2023
2      16/02/2023
3      09/01/2013
4      22/07/2011
          ...    
259    09/03/2007
260    27/12/2010
261    03/01/2012
262    12/03/2009
263    30/10/2008
Name: data_status, Length: 264, dtype: object

In [11]:
# 4) Seu código aqui:

anos_requisicoes_completos = datas_completas.str.split('/')
anos_requisicoes_ano = anos_requisicoes_completos.str.slice(start = -1)

anos_requisicoes_ano.value_counts()

data_status
[2008]    37
[2007]    29
[2023]    28
[2014]    20
[2010]    19
[2019]    16
[2013]    15
[2020]    15
[2012]    13
[2022]    11
[2011]    10
[2009]     9
[2015]     8
[2024]     6
[2017]     5
[2021]     5
[2016]     4
[2018]     4
[2002]     4
[2006]     4
[2003]     1
[2004]     1
Name: count, dtype: int64

####  2) Melhorando a interação com o API
Observe dois URLs de consultas diferentes, por exemplo o URL utilizado em aula, e o URL feito no exercício anterior. Compare-os e observe as diferenças.

1. Faça uma função em Python que recebe como argumento o UF da consulta e o tipo de interessado (```'Estado'```ou ```Município```), e que devolve os dados da consulta no formato *DataFrame*.
2. Quantas solicitações para o Estado podem ser consultadas para Minas Gerais com *status* em 'Arquivado por decurso de prazo' estão registradas?
3. Qual é o município da Bahia com mais solicitações deferidas?
4. Salve um arquivo .csv com os dados de solicitações da Bahia, com interessado = 'Estado'

In [12]:
#1) Seu código aqui
def request_uf_tipo(uf: str, tipo_interessado: str):
    resposta = requests.get('https://apidatalake.tesouro.gov.br/ords/sadipem/tt/pvl')
    dados = resposta.json()
    df = pd.DataFrame(dados['items'])
    
    selecao = (df['uf'] == uf) & (df['tipo_interessado'] == tipo_interessado)
    df_selecao = df[selecao]
    
    return df_selecao

In [13]:
# 2) Seu código aqui
uf = 'MG'
tipo_interessado = 'Estado'

df_selecao = request_uf_tipo(uf, tipo_interessado)
df_selecao.head(5)

id_pleito tipo_interessado   interessado  cod_ibge  uf  \
1531      13672           Estado  Minas Gerais        31  MG   
1586      13587           Estado  Minas Gerais        31  MG   
2262      13695           Estado  Minas Gerais        31  MG   
2327      20458           Estado  Minas Gerais        31  MG   
2953       9197           Estado  Minas Gerais        31  MG   

                   num_pvl                                             status  \
1531                  None  Encaminhado à PGFN com manifestação técnica fa...   
1586                  None  Encaminhado à PGFN com manifestação técnica fa...   
2262                  None                                          Arquivado   
2327  00000.000000/2020-68                                 Arquivado a pedido   
2953                  None  Encaminhado à PGFN com manifestação técnica fa...   

              num_processo        data_protocolo  \
1531  19405.000048/2005-14  2005-12-28T02:00:00Z   
1586  19405.000031/2004-78  2004-11-27T02:00:00Z   
2262  19405.000053/2003-57  2003-08-25T03:00:00Z   
2327  17944.001085/2015-81  2015-11-23T02:00:00Z   
2953  17944.000924/2009-03  2010-03-10T03:00:00Z   

                                          tipo_operacao  \
1531  Operação contratual externa (com garantia da U...   
1586  Operação contratual externa (com garantia da U...   
2262  Operação contratual externa (com garantia da U...   
2327  Operação contratual externa (com garantia da U...   
2953  Operação contratual externa (com garantia da U...   

                 finalidade                           tipo_credor  \
1531         Multissetorial  Instituição Financeira Internacional   
1586  Desenvolvimento rural  Instituição Financeira Internacional   
2262         Infraestrutura  Instituição Financeira Internacional   
2327         Multissetorial  Instituição Financeira Internacional   
2953         Infraestrutura  Instituição Financeira Internacional   

                                                 credor          moeda  \
1531  Banco Internacional para Reconstrução e Desenv...  Dólar dos EUA   
1586  Banco Internacional para Reconstrução e Desenv...  Dólar dos EUA   
2262  Banco Internacional para Reconstrução e Desenv...  Dólar dos EUA   
2327            Banco Interamericano de Desenvolvimento  Dólar dos EUA   
2953            Banco Interamericano de Desenvolvimento  Dólar dos EUA   

            valor  pvl_assoc_divida  pvl_contradado_credor data_status  
1531  170000000.0                 1                      0  19/04/2006  
1586   35000000.0                 1                      0  09/12/2004  
2262   35000000.0                 0                      0  20/02/2004  
2327   50000000.0                 0                      0  29/04/2016  
2953   50000000.0                 1                      0  01/04/2010

In [14]:
status = df_selecao[df_selecao['status'] == 'Arquivado por decurso de prazo'].value_counts()
status

Series([], Name: count, dtype: int64)

## Como a consulta mostra, não há nenhum status do tipo 'Arquivado por decurso de prazo'

In [16]:
# 3) Seu código aqui
uf = 'BA'
tipo_interessado = 'Município'

df_selecao = request_uf_tipo(uf, tipo_interessado)
df_selecao.head(5)

id_pleito tipo_interessado    interessado  cod_ibge  uf  \
18       67739        Município       Juazeiro   2918407  BA   
21       65397        Município    Nova Fátima   2922730  BA   
26       65857        Município        Lajedão   2918902  BA   
91       65993        Município  Santo Estêvão   2928802  BA   
129      65435        Município   Porto Seguro   2925303  BA   

                  num_pvl                           status  \
18   PVL02.004332/2023-48  Em retificação pelo interessado   
21   PVL02.002486/2023-03                Deferido (PVL-IF)   
26   PVL02.002869/2023-73                Deferido (PVL-IF)   
91   PVL02.002767/2023-58                         Deferido   
129  PVL02.002323/2023-12                         Deferido   

             num_processo        data_protocolo  \
18   17944.000222/2024-51  2024-01-11T18:34:19Z   
21                   None                  None   
26                   None                  None   
91   17944.103592/2023-68  2023-08-07T20:57:39Z   
129  17944.102783/2023-11  2023-07-17T18:35:14Z   

                                         tipo_operacao      finalidade  \
18   Operação contratual interna (com garantia da U...  Infraestrutura   
21                         Operação contratual interna  Infraestrutura   
26                         Operação contratual interna  Infraestrutura   
91                         Operação contratual interna  Infraestrutura   
129                        Operação contratual interna  Infraestrutura   

                         tipo_credor  \
18   Instituição Financeira Nacional   
21   Instituição Financeira Nacional   
26   Instituição Financeira Nacional   
91   Instituição Financeira Nacional   
129  Instituição Financeira Nacional   

                                        credor moeda        valor  \
18                         Banco do Brasil S/A  Real  100000000.0   
21   Agência de Fomento do Estado da Bahia S/A  Real    2000000.0   
26   Agência de Fomento do Estado da Bahia S/A  Real    1500000.0   
91                     Caixa Econômica Federal  Real   16000000.0   
129                    Caixa Econômica Federal  Real  100000000.0   

     pvl_assoc_divida  pvl_contradado_credor data_status  
18                  0                      0  25/01/2024  
21                  0                      1  30/06/2023  
26                  0                      1  28/07/2023  
91                  0                      1  14/08/2023  
129                 0                      1  20/07/2023

In [17]:
df_selecao_deferido = df_selecao[df_selecao['status'] == 'Deferido']
municipios = df_selecao_deferido['interessado'].value_counts()
print(municipios)

interessado
Luís Eduardo Magalhães    4
Barreiras                 3
Mucuri                    3
Teixeira de Freitas       3
Santo Estêvão             3
                         ..
Ibicuí                    1
Porto Seguro              1
Feira da Mata             1
Encruzilhada              1
Malhada                   1
Name: count, Length: 68, dtype: int64


In [18]:
mun_maisSolic = municipios.head(1)

print(mun_maisSolic)

interessado
Luís Eduardo Magalhães    4
Name: count, dtype: int64


### O município da Bahia com mais solicitações 'Deferidas' é Luís Eduardo Magalhães, com 4 situações desse tipo.

In [19]:
# 4) Seu código aqui

uf = 'BA'
tipo_interessado = 'Estado'

df_selecao = request_uf_tipo(uf, tipo_interessado)
df_selecao.head(5)

id_pleito tipo_interessado interessado  cod_ibge  uf  \
4         60782           Estado       Bahia        29  BA   
1321       8858           Estado       Bahia        29  BA   
1369      64646           Estado       Bahia        29  BA   
2096      12327           Estado       Bahia        29  BA   
2318      18531           Estado       Bahia        29  BA   

                   num_pvl                                             status  \
4     PVL02.007004/2022-12                                           Deferido   
1321                  None  Encaminhado à PGFN com manifestação técnica fa...   
1369  PVL02.000882/2023-98                                         Em análise   
2096                  None  Encaminhado à PGFN com manifestação técnica fa...   
2318                  None                                           Deferido   

              num_processo        data_protocolo  \
4     17944.102769/2022-28  2022-07-14T13:15:21Z   
1321  17944.000853/2014-06  2014-07-10T03:00:00Z   
1369  17944.102176/2023-42  2024-06-26T19:13:54Z   
2096  17944.001659/2008-91  2009-03-24T03:00:00Z   
2318  19407.000044/2002-65  2002-06-11T03:00:00Z   

                                          tipo_operacao  \
4                           Operação contratual interna   
1321  Operação contratual externa (com garantia da U...   
1369  Operação contratual externa (com garantia da U...   
2096  Operação contratual externa (com garantia da U...   
2318                        Operação contratual interna   

                        finalidade                           tipo_credor  \
4                   Infraestrutura       Instituição Financeira Nacional   
1321                Infraestrutura  Instituição Financeira Internacional   
1369                Infraestrutura  Instituição Financeira Internacional   
2096  Fortalecimento Institucional  Instituição Financeira Internacional   
2318             Saneamento básico       Instituição Financeira Nacional   

                                                 credor          moeda  \
4                                 Banco de Brasília S/A           Real   
1321  Banco Internacional para Reconstrução e Desenv...  Dólar dos EUA   
1369                       Corporação Andina de Fomento  Dólar dos EUA   
2096            Banco Interamericano de Desenvolvimento  Dólar dos EUA   
2318                            Caixa Econômica Federal           Real   

             valor  pvl_assoc_divida  pvl_contradado_credor data_status  
4     2.000000e+08                 0                      0  15/07/2022  
1321  4.000000e+08                 1                      0  16/07/2014  
1369  1.500000e+08                 0                      0  08/07/2024  
2096  4.090000e+08                 1                      0  27/03/2009  
2318  7.360877e+06                 1                      0  05/07/2002

In [22]:
df_selecao.to_csv('df_selecao.csv')
df_selecao.head()

id_pleito tipo_interessado interessado  cod_ibge  uf  \
4         60782           Estado       Bahia        29  BA   
1321       8858           Estado       Bahia        29  BA   
1369      64646           Estado       Bahia        29  BA   
2096      12327           Estado       Bahia        29  BA   
2318      18531           Estado       Bahia        29  BA   

                   num_pvl                                             status  \
4     PVL02.007004/2022-12                                           Deferido   
1321                  None  Encaminhado à PGFN com manifestação técnica fa...   
1369  PVL02.000882/2023-98                                         Em análise   
2096                  None  Encaminhado à PGFN com manifestação técnica fa...   
2318                  None                                           Deferido   

              num_processo        data_protocolo  \
4     17944.102769/2022-28  2022-07-14T13:15:21Z   
1321  17944.000853/2014-06  2014-07-10T03:00:00Z   
1369  17944.102176/2023-42  2024-06-26T19:13:54Z   
2096  17944.001659/2008-91  2009-03-24T03:00:00Z   
2318  19407.000044/2002-65  2002-06-11T03:00:00Z   

                                          tipo_operacao  \
4                           Operação contratual interna   
1321  Operação contratual externa (com garantia da U...   
1369  Operação contratual externa (com garantia da U...   
2096  Operação contratual externa (com garantia da U...   
2318                        Operação contratual interna   

                        finalidade                           tipo_credor  \
4                   Infraestrutura       Instituição Financeira Nacional   
1321                Infraestrutura  Instituição Financeira Internacional   
1369                Infraestrutura  Instituição Financeira Internacional   
2096  Fortalecimento Institucional  Instituição Financeira Internacional   
2318             Saneamento básico       Instituição Financeira Nacional   

                                                 credor          moeda  \
4                                 Banco de Brasília S/A           Real   
1321  Banco Internacional para Reconstrução e Desenv...  Dólar dos EUA   
1369                       Corporação Andina de Fomento  Dólar dos EUA   
2096            Banco Interamericano de Desenvolvimento  Dólar dos EUA   
2318                            Caixa Econômica Federal           Real   

             valor  pvl_assoc_divida  pvl_contradado_credor data_status  
4     2.000000e+08                 0                      0  15/07/2022  
1321  4.000000e+08                 1                      0  16/07/2014  
1369  1.500000e+08                 0                      0  08/07/2024  
2096  4.090000e+08                 1                      0  27/03/2009  
2318  7.360877e+06                 1                      0  05/07/2002